### PR TITLE
Authz v2 tests

### PR DESCRIPTION
**Pull request by bhaskaramVaranasi**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: test/integration/authz_v2_test.go [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+// Output: AP updated to match with existing AD
+func TestUpdatedAuthorizationPolicyRestoresOriginal(t *testing.T) {
+	// Basic setup
+	// Creates Athenz Domain, Service verifies Authorization Policy
+	e := fixtures.GetBasicRbacV2Case(nil)
+	rolloutAndValidateAuthorizationPolicyScenario(t, e, create, create)
+
+	// Retrieve a modified authorization policy
+	modifiedResources := fixtures.GetBasicRbacV2Case(&fixtures.RbacV2Modifications{
+		ModifyAthenzDomain: []func(sginedDomain *zms.SignedDomain){},
```
Comment (kpharasi) [inline test/integration/authz_v2_test.go:149]: typo

File: test/integration/authz_v2_test.go [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+// Initial: Existing service with annotation, AP
+// Input Actions: delete service with annotation
+// Output: AP deleted
+func TestDeleteServiceShouldDeleteAuthorizationPolicy(t *testing.T) {
+	// Initial set up
+	e := fixtures.GetBasicRbacV2Case(nil)
+	rolloutAndValidateAuthorizationPolicyScenario(t, e, create, create)
+
+	// Delete athenz Domain
```
Comment (kpharasi) [inline test/integration/authz_v2_test.go:268]: Delete service?

File: test/integration/authz_v2_test.go [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+// createServiceObjects - Creates the list of kubernetes service objects as passed in
+func createServiceObjects(t *testing.T, services []*v1.Service) {
+	for _, service := range services {
+		_, err := framework.Global.K8sClientset.CoreV1().Services(service.Namespace).Create(service)
+		assert.Nil(t, err, "service create error should be nil")
+	}
+}
+
+// updateServiceObjects - Updates the list of Kubernetes serivce objects as passed in
```
[the remark this answers is not in the document]
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go]: fixed it

File: test/integration/authz_v2_test.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+	}
+}
+
+// deleteServiceObjects - Delete the list of kubernetes services passed in
+func deleteServiceObjects(t *testing.T, services []*v1.Service) {
+	for _, service := range services {
+		err := framework.Global.K8sClientset.CoreV1().Services(service.Namespace).Delete(service.Name, &metav1.DeleteOptions{})
+		assert.Nil(t, err, "service delete error should be nil")
+	}
+	services = []*v1.Service{}
```
[the remark this answers is not in the document]
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go]: Not necessary removed

File: test/integration/authz_v2_test.go (hunk 1 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go]: fixed it.

File: test/integration/authz_v2_test.go (hunk 2 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go]: Updated the comment.

File: test/integration/authz_v2_test.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+	case delete:
+		err := framework.Global.AthenzDomainClientset.AthenzV1().AthenzDomains().Delete(athenzDomain.Name, metav1.DeleteOptions{})
+		assert.Nil(t, err, "athenz domain delete error should be nil")
+	}
+}
+
+// applyServices Based on the action specified creates, updates or deletes services
+func applyServices(t *testing.T, services []*v1.Service, a action) {
+	switch a {
+	case create:
```
Comment (RuiGuo34) [inline test/integration/authz_v2_test.go:83]: can you make applyServices and applyAthenzDomain in the same format? Either have helper function call in each case or just include the code for each case

File: test/integration/authz_v2_test.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+// Initial: No services, existing AD
+// Input Actions: create service with annotation
+// Output: AP created
+func TestCreateAuthorizationPolicy(t *testing.T) {
+	e := fixtures.GetBasicRbacV2Case(nil)
+	rolloutAndValidateAuthorizationPolicyScenario(t, e, create, create)
+	cleanupAuthorizationRbac(t, e)
+}
+
+// Manual update AP
```
Comment (RuiGuo34) [inline test/integration/authz_v2_test.go:136]: possible to add a case which override annotaion is set for an ap?

File: test/integration/authz_v2_test.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+	ap := e.AuthorizationPolicies[0]
+	err := framework.Global.IstioClientset.Delete(ap.GroupVersionKind(), ap.Name, ap.Namespace)
+	assert.Nil(t, err, "Delete of authorization policy should not fail")
+
+	// Verification and cleanup
+	rolloutAndValidateAuthorizationPolicyScenario(t, e, noop, noop)
+	cleanupAuthorizationRbac(t, e)
+}
+
+// Update AP
```
Comment (RuiGuo34) [inline test/integration/authz_v2_test.go:192]: Can you rename this test case to update AD?

File: test/integration/authz_v2_test.go [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+// Output: after member expired, AP update with this member removal
+func TestUpdateAuthorizationPolicyRemovingExpiredMembers(t *testing.T) {
+	// Initial set up
+	e := fixtures.GetBasicRbacV2Case(nil)
+	rolloutAndValidateAuthorizationPolicyScenario(t, e, create, create)
+
+	// Update Athenz Domain with an expiring member
+	// Get updated fixtures
+	newUserName := "user.bar"
+	value := time.Now().Add(time.Second * time.Duration(30))
```
Comment (RuiGuo34) [inline test/integration/authz_v2_test.go:332]: feel like there is no need to add 30s for current time, we can just set the expiration time to current time, and reduce the time to wait for expiration to a shorter period

File: test/integration/authz_v2_test.go (hunk 5 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go:99]: Done

File: test/integration/authz_v2_test.go (hunk 6 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go:152]: Added testcase

File: test/integration/authz_v2_test.go (hunk 7 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go]: Done, updated the comment .

File: test/integration/authz_v2_test.go (hunk 8 above)
Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go:392]: We poll for the authorization policies every second for 5 times (seconds). If the expiry time is set to time now, we cannot be sure if the update for the Authorization Policy did not go through or if the member role expired before the athenz domain is processed when this test case fails. I tested out various delays and found that 30seconds and 1 minute reliably works.

**Review by kpharasi — APPROVED** [review]
👍

File: test/integration/authz_v2_test.go (hunk 8 above)
Comment (RuiGuo34) [reply test/integration/authz_v2_test.go:392]: ok got it.

File: test/integration/authz_v2_test.go [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+
+	rolloutAndValidateAuthorizationPolicyScenario(t, modified, update, noop)
+	cleanupAuthorizationRbac(t, modified)
+}
+
+// Update AD
+// Initial: Existing service with annotation, AP
+// Input Actions: update athenz domain with service related role, or policies associated to the role
+// Output: AP updated
+func TestUpdateAthenzDomainIgnoresAuthorizationPolicyWithOverrideAnnotation(t *testing.T) {
```
Comment (RuiGuo34) [inline test/integration/authz_v2_test.go:251]: I think the test case should be: existing ap with override annotation set, input actions: update ap manually with policy addition, output: ap is update(does not sync with ad)
lmk if this makes sense

Comment (bhaskaramVaranasi) [reply test/integration/authz_v2_test.go:251]: Yup, makes sense. Updated the comment with the suggested text.

Comment (RuiGuo34) [reply test/integration/authz_v2_test.go:251]: thank you

**Review by RuiGuo34 — APPROVED** [review]

**Pull request merged**